library/math/dual.lisp

```diff
@@ -0,0 +1,177 @@
+;;;; dual.lisp
+
+
+(coalton-library/utils:defstdlib-package #:coalton-library/math/dual
+  (:use
+   #:coalton
+   #:coalton-library/builtin
+   #:coalton-library/classes
+   #:coalton-library/functions
+   #:coalton-library/math/arith
+   #:coalton-library/math/elementary
+   #:coalton-library/math/integral
+   #:coalton-library/hash)
+  (:export
+   #:Dual
+   #:primal-part
+   #:dual-part)
+  (:documentation "
+dual numbers are a hypercomplex number system [1]. A dual number has the form
+a + bε where a and b are real numbers and ε is a symbol that satisfies ε^2=0
+and ε!=0. One application of dual numbers is automatic differentiation; an example
+taken from [2] is as follows:
+
+    consider you have the given expression f(x) = 3x+2 and you want to calculate
+    f(4) and f'(4).
+
+    You then convert 4 into a dual number which gives:
+
+        4+1ε
+
+    then the computation is as follows:
+
```

Comment:
This math isn't stated very clearly, because you've written:

$$f(x)=3x+2$$

as

$$f(x) = (3+0\varepsilon)x+(2+0\varepsilon).$$

This isn't bad in and of itself, but the exposition could be clearer, or this could be eliminated altogether in the following way:

We have $f(x)=3x+2$. So:
$$f(4+\varepsilon) = 3(4+\varepsilon)+2 = 12 + 3\varepsilon+2 = 14+3\varepsilon.$$

In this case, we don't need to make use of the fact that $\varepsilon^2 = 0$.

So to move forward, I'd recommend one of three things:

- Option 1: Keep the current exposition, but be clear that you are now evaluating the function $f(x) = (3+0\varepsilon)x+(2+0\varepsilon)$.
- Option 2: Change the current exposition to use the simplified approach to evaluating $f(x) = 3x+2$.
- Option 3: (My preference:) Use an example like $f(x) = 3x^2-1$ instead, with the simpler style of exposition, just so we can see the $\varepsilon^2$ law in action.

Choosing any one of these three options is required for this PR.